user/tests/wiring/pwm/pwm.cpp

```diff
@@ -0,0 +1,400 @@
+}
+
+template <typename F> void for_all_pwm_pins(F callback, hal_pin_t skipPin = PIN_INVALID)
+{
+    RGB.control(true);        
+    for (uint8_t i = 0; i<arraySize(pwm_pins); i++)
+    {
+        if (skipPin != PIN_INVALID && pwm_pins[i].pin == skipPin) {
+            continue;
+        }
+        callback(pwm_pins[i].pin, pwm_pins[i].name);
+    }
+    RGB.control(false);
+}
+
+test(PWM_01_LowDCAnalogWriteOnPinResultsInCorrectPulseWidth) {
+    for_all_pwm_pins([](hal_pin_t pin, const char* name) {
+        pwm_output_pin = pin;
+        pinMode(pwm_output_pin, OUTPUT);
+        //out->printlnf("Output pin: %s", name);
+
+        for_all_pwm_pins([](hal_pin_t pin, const char* name) {
+
+            //out->printlnf("Input pin: %s", name);
+            pinMode(pin, INPUT);
+            uint32_t avgPulseHigh = 0;
+
+            // 8-bit resolution
+            analogWriteResolution(pwm_output_pin, 8);
+            assertEqual(analogWriteResolution(pwm_output_pin), 8);
+            // 9.8% Duty Cycle at 500Hz = 196us HIGH, 1804us LOW.
+            analogWrite(pwm_output_pin, 25);
+            avgPulseHigh = get_average_pulse_width(pin);
+            // avgPulseHigh should equal 200 +/- 50
+            assertMoreOrEqual(avgPulseHigh, 150);
+            assertLessOrEqual(avgPulseHigh, 250);
+
+            // 4-bit resolution
+            analogWriteResolution(pwm_output_pin, 4);
+            assertEqual(analogWriteResolution(pwm_output_pin), 4);
+            // 13.3% Duty Cycle at 500Hz = 266us HIGH, 1733us LOW.
+            analogWrite(pwm_output_pin, 2);
+            avgPulseHigh = get_average_pulse_width(pin);
+            // avgPulseHigh should equal 266 +/- 50
+            assertMoreOrEqual(avgPulseHigh, 216);
+            assertLessOrEqual(avgPulseHigh, 316);
+
+            // 12-bit resolution
+            analogWriteResolution(pwm_output_pin, 12);
+            assertEqual(analogWriteResolution(pwm_output_pin), 12);
+            // 10% Duty Cycle at 500Hz = 200us HIGH, 1800us LOW.
+            analogWrite(pwm_output_pin, 409);
+            avgPulseHigh = get_average_pulse_width(pin);
+            // avgPulseHigh should equal 200 +/- 50
+            assertMoreOrEqual(avgPulseHigh, 150);
+            assertLessOrEqual(avgPulseHigh, 250);
+
+            // 15-bit resolution
+            analogWriteResolution(pwm_output_pin, 15);
+            assertEqual(analogWriteResolution(pwm_output_pin), 15);
+            // 10% Duty Cycle at 500Hz = 200us HIGH, 1800us LOW.
+            analogWrite(pwm_output_pin, 3277);
+            avgPulseHigh = get_average_pulse_width(pin);
+            // avgPulseHigh should equal 200 +/- 50
+            assertMoreOrEqual(avgPulseHigh, 150);
+            assertLessOrEqual(avgPulseHigh, 250);
+        }, pwm_output_pin);
+        pinMode(pwm_output_pin, INPUT);
+    });
+}
+
+// 10hz
+test(PWM_02_LowFrequencyAnalogWriteOnPinResultsInCorrectPulseWidth) {
+    static const uint32_t TEST_FREQUENCY_HZ = 10;
```

Comment:
Since we guarantee operation down to 5Hz, should we make this test run at 5Hz?  I did a quick test locally doubling the assert values and it passed.

Reply:
Probably? This is just what was in `no_fixture_long_running`

Reply:
I would advocate for changing it to 5Hz in `no_fixture_long_running` as well (maybe as a separate PR later).  If we going to test it, might as well test the limits?

Reply:
[I agree with that, made a story for it here](https://app.shortcut.com/particle/story/108164/update-pwm-tests-to-use-lowest-supported-frequency)